### DISCUSSION TITLE
Is the field `hidden_meths` in class tables ever used ?

### DISCUSSION THREAD

**Discussion by lthls** [ocaml/ocaml#13236] [General]
I am again trying to understand the code for objects and classes, and I cannot find any code that triggers non-trivial use of the `hidden_meths` field in `CamlinternalOO.table` (i.e. in my experiments, it's always empty).
The `methods_by_label` seems to be linked, as its only use is to decide when methods go into the `hidden_meths` list instead of the method array. I believe that this tracks, for each method, whether it is virtual or concrete in the current context.

My suspicion is that these fields were useful for a short period of time, when virtual classes also had tables, but now that virtual classes only contain a `class_init` function to update a table during inheritance, actual tables are only created for concrete classes and do not need to track virtual methods anymore.

But I am not very confident in my analysis, and would welcome either confirmation, or a counter-example (a test program that requires storing hidden methods), or just some hints.

I'm assuming that the people most likely to have an answer are @garrigue and @vouillon, although if anyone else has some insight on this feel free to comment.

**Comment by lthls** [comment]
I eventually found an example that triggers the use of the field:
```ocaml
class virtual c_v : object method virtual c : int end = object method c = 0 end

class c_c : object method c : int end = object (self)
  inherit c_v
  method c = 42
end
```
I am still trying to understand what exactly is going on, so any insights are welcome.

**Comment by shindere** [comment]
If it was so hard to find such an example maybe it's worth writing a
test to be added to the testsuite to make sure that part of the compiler
is exercized?

**Reply by lthls** [reply]
Yes, I will propose adding more tests, but later.
